.github/workflows/yocto-build-deploy.yml

```
@@ -0,0 +1,1174 @@
+      # Example output: JSON_ARRAY='["qemu"]'
+      # Required to create test suite matrix
+      - id: test_workers
+        name: Prerequiste - Create a JSON array for test_workers
+        shell: bash
+        run: |
+          set -x
+
+          # Remove spaces and newlines from the input
+          INPUT="$(echo "${INPUT}" | tr -d '[:space:]')"
+
+          # Convert to JSON array using jq to split the string by the given delimiter
+          # Add a step to check if the array is empty, default to [""] if so
+          JSON_ARRAY="$(jq --raw-input --compact-output --null-input --arg delim "$DELIMITER" --arg input "$INPUT" '
+            if $input == "" then
+              [""] # Default to a single empty string element if input is empty
+            else
+              $input | split($delim) # Split the input by the delimiter
+            end
+          ')"
+
+          echo "build=${JSON_ARRAY}" >> "${GITHUB_OUTPUT}"
+          echo "json=${JSON_ARRAY}" >> "${GITHUB_OUTPUT}"
+        env:
+          INPUT: ${{ inputs.test-workers }}
+          DELIMITER: ","
+
+      # Creates an array for the test suites
+      # Example output: JSON_ARRAY='["os","hup","cloud"]'
+      # Required to create test suite matrix
+      - id: test_suites
+        name: Prerequiste - Create a JSON array for test_suites
+        shell: bash
+        run: |
+          set -x
+
+          # Remove spaces and newlines from the input
+          INPUT="$(echo "${INPUT}" | tr -d '[:space:]')"
+
+          # Convert to JSON array using jq to split the string by the given delimiter
+          # Add a step to check if the array is empty, default to [""] if so
+          JSON_ARRAY="$(jq --raw-input --compact-output --null-input --arg delim "$DELIMITER" --arg input "$INPUT" '
+            if $input == "" then
+              [""] # Default to a single empty string element if input is empty
+            else
+              $input | split($delim) # Split the input by the delimiter
+            end
+          ')"
+
+          echo "build=${JSON_ARRAY}" >> "${GITHUB_OUTPUT}"
+          echo "json=${JSON_ARRAY}" >> "${GITHUB_OUTPUT}"
+        env:
+          INPUT: ${{ inputs.test-suites }}
+          DELIMITER: ","
+
+      # Creates a test matrix to test a number of suites
+      # We do this to run each suite in parallel if there are available workers, and to be able to retry individual test suites
+      # Example output: + json='{"TEST_SUITE":["os","hup","cloud"],"DEVICE_TYPE":["generic-amd64"],"ENVIRONEMENT":["balenaos-balenamachine"],"WORKER_TYPE":["qemu"]}'
+      - name: Create Leviathan test matrix
+        id: leviathan_test_matrix
+        if: inputs.run-tests == true
+        env:
+          MATRIX: >
+            {
+              "TEST_SUITE": ${{ steps.test_suites.outputs.json }},
+              "DEVICE_TYPE": ${{ format('["{0}"]', steps.balena-lib.outputs.device_slug) }},
+              "ENVIRONEMENT": ${{ format('["{0}"]', inputs.test-environment) }},
```

This one is also puzzling me a bit. I believe this is a typo, but:

1. The example output above also says `ENVIRONEMENT` like here (possibly just copy and paste).
2. I couldn't spot where this is used to double check it.

```suggestion
              "ENVIRONMENT": ${{ format('["{0}"]', inputs.test-environment) }},
```

---

fixed!

---

Hi, thanks that's my english clearly failing me sometimes. I misspell environment all the time with that extra E. 
It happen sometimes when I am writing `levaithan` too much instead of `leviathan`